### PR TITLE
Add support for Hashicorp Vault to Viper

### PR DESCRIPTION
**Pull request by netik**
About 17 people were asking for this feature, so here goes...

This supports Hashicorp vault as a secure configuration backend, either by using the VAULT_TOKEN environment variable as a secret or using approle authentication using VAULT_SECRET_ID and VAULT_ROLE_ID. 

I have tested this code against my vault server but didn't want to introduce a test dependency on vault. I'm willing to set up vault in docker and write tests if required.

### CODE REVIEW

**Comment by kevholditch** [comment]
Awesome work @netik, I was about to look into writing this feature as I need it for a project.  Thought I would check the PRs first...  Contribs any idea when this will be merged?

**Comment by netik** [comment]
I don't have any control over that, but if you need help making it go,
please let me know and I'll help!

-john

**Comment by kevholditch** [comment]
thanks @netik appreciate that, think I'll wait for this to be merged then it makes consuming it easier.  Thanks again for doing this work, really awesome

**Comment by kevholditch** [comment]
@spf13 is this something you are planning to merge into the `master` branch?  viper is a great library and having vault support would be awesome.

**Comment by netik** [comment]
Just pinging on this issue, I'd love to merge this in if no one objects.

**Comment by eliastor** [comment]
Any news?

File: vault/vault.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+ * export VAULT_SECRET_ID= ... secret ...
+ * export VAULT_ROLE_ID= ... role id ...
+ * -- or --
+ * export VAULT_TOKEN = ....
+ *
+ * If you are using SSL with vault, you can set
+ * export VAULT_CACERT= ... pem file containing ca cert ...
+ *           and/or
+ * export VAULT_SSL_VERIFY=no
+ */
```
Comment (adamdecaf) [inline vault/vault.go]: To get a package level doc this should be:

```Go
/**
  Package vault implements...
 */
package vault
```

**Comment by denouche** [comment]
Hello, any news about this PR?

**Comment by Strum355** [comment]
@netik any chance you could rebase upstream master into your branch? no idea when this is gonna be merged and id like to have at least *something* to work with even if its not from upstream :)

**Comment by sagikazarmark** [comment]
Until this (or a similar PR) gets merged, here is a Viper extension with Vault support:

https://github.com/sagikazarmark/viperx

You can see it in action here: https://github.com/banzaicloud/cloudinfo/pull/229

**Comment by keshavmaheshwari** [comment]
When can this PR be merged?

**Comment by sagikazarmark** [comment]
Personally, I'm not a huge fan of how remote key-value stores are implemented in Viper. So I would rather rethink the foundations before merging any more providers.

In the meantime, you can check the project I linked above: https://github.com/sagikazarmark/viperx

It adds support for Vault to Viper.

**Comment by quakelee** [comment]
Any update on this PR?

**Comment by jtbonhomme** [comment]
Hello, any update on this PR ?
Cheers !

**Comment by wwojcik** [comment]
Any news?

**Comment by sagikazarmark** [comment]
Remote providers will probably be changed in the future, so Vault as a provider will probably not land in the near future.

You can try using https://github.com/sagikazarmark/viperx
We use it in production at Banzai Cloud

**Comment by zephinzer** [comment]
This would be really useful @spf13

**Comment by astromahi** [comment]
Do we have any news on this @netik ?. Viper will be a complete config manager if vault has been integrated. Due to this reason I have to roll my own library to support vault.

@spf13

**Comment by sagikazarmark** [comment]
**tl;dr** Viper can be very easily extended to support Vault, here is an example:

https://github.com/sagikazarmark/viperx

---

As an effort to make Viper better we are (well, at least I am) now working on Viper 2.

Unfortunately, Viper already has a large amount of external dependencies that pulls in lots of unnecessary stuff for a lot of people. In order to not add more dependencies, I think this PR should not be merged until #889 is sorted out.

Since I also use Viper with Vault, I created a library that works quite well with them:

https://github.com/sagikazarmark/viperx

You can give it a go (or create your own). Viper can be extended with remote providers quite easily, so I don't think this is something that Viper needs to support OOTB right now.

Chances are this won't be merged until #889 is sorted out and the relevant decisions are made....at least not by me.
